fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesCostModel.java

```diff
@@ -223,6 +236,47 @@ private OptionalInt comparePrimaryScanToIndexScan(@Nonnull Map<Class<? extends R
         return OptionalInt.empty();
     }
 
+    private OptionalInt compareInUnion(@Nonnull final RelationalExpression leftExpression,
+                                       @Nonnull final RelationalExpression rightExpression) {
+        if (!(leftExpression instanceof RecordQueryInUnionPlan)) {
+            return OptionalInt.empty();
+        }
+
+        //
+        // If both an InUnions we just return 0
+        //
+        if (rightExpression instanceof RecordQueryInUnionPlan) {
+            return OptionalInt.of(0);
+        }
+
+        final RecordQueryInUnionPlan inUnionPlan = (RecordQueryInUnionPlan)leftExpression;
+
+        // right is not in union
+
+        // If at least one scan comparison on the in union side uses a comparison to the in-values, then the in union
+        // plan is useful.
+        final Set<ScanComparisons> scanComparisonsSet = ScanComparisonsProperty.evaluate(inUnionPlan);
+
+        final ImmutableSet<String> parametersInScanComparisons =
+                scanComparisonsSet
+                        .stream()
+                        .flatMap(scanComparisons ->
+                                scanComparisons.getEqualityComparisons()
+                                        .stream()
+                                        .filter(comparison -> comparison instanceof Comparisons.ParameterComparison)
+                                        .map(comparison -> (Comparisons.ParameterComparison)comparison))
+                        .map(Comparisons.ParameterComparison::getParameter)
+                        .collect(ImmutableSet.toImmutableSet());
+
+        if (inUnionPlan.getValuesSources()
+                .stream()
+                .noneMatch(inValuesSource -> parametersInScanComparisons.contains(inValuesSource.getBindingName()))) {
+            return OptionalInt.of(1);
```

Comment:
Probably needs a comment here, too. This evidently means that the right is "better," so that the left can be removed from consideration.

It might be better to use `anyMatch` and reverse the `1` and `0`, to have this code match the sense of the above comment, "at least one ..."

Reply:
I changed the comment as I think we only want to exclude the bad in union plan here, no necessary make the in union plan win. So now thing should be in-sync.